tests/Doctrine/Tests/ORM/Functional/Ticket/GH7836Test.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\ORM\PersistentCollection;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use function assert;
+
+/**
+ * @group GH7836
+ */
+class GH7836Test extends OrmFunctionalTestCase
+{
+    protected function setUp() : void
+    {
+        parent::setUp();
+
+        $this->setUpEntitySchema([GH7836ParentEntity::class, GH7836ChildEntity::class]);
+
+        $parent = new GH7836ParentEntity();
+        $parent->addChild(100, 'foo');
+        $parent->addChild(100, 'bar');
+        $parent->addChild(200, 'baz');
+
+        $this->_em->persist($parent);
+        $this->_em->flush();
+        $this->_em->clear();
+    }
+
+    public function testMatchingRespectsCollectionOrdering() : void
+    {
+        $parent = $this->_em->find(GH7836ParentEntity::class, 1);
+        assert($parent instanceof GH7836ParentEntity);
+
+        $children = $parent->getChildren()->matching(Criteria::create());
+
+        self::assertEquals(100, $children[0]->position);
```

Comment:
Please use `self::assertSame()` to be a bit more strict (yay strictness :metal:)

Reply:
done!